libs/api/ebpf_api.cpp

```diff
@@ -1948,6 +2144,267 @@ ebpf_program_unload(_In_ struct bpf_program* program)
+            service_name.c_str(), get_wstring_from_string(file_name_string).c_str(), &service_handle);
+        if (error != ERROR_SUCCESS) {
+            result = win32_error_code_to_ebpf_result(error);
+            goto Done;
+        }
+
+        // Create registry path and update module ID in the service path.
+        paramaters_path = paramaters_path + service_name.c_str() + L"\\" + SERVICE_PARAMETERS;
+        error = _ebpf_create_registry_key(HKEY_LOCAL_MACHINE, paramaters_path.c_str());
+        if (error != ERROR_SUCCESS) {
+            result = win32_error_code_to_ebpf_result(error);
+            goto Done;
+        }
+        error = _ebpf_update_registry_value(
+            HKEY_LOCAL_MACHINE, paramaters_path.c_str(), REG_BINARY, NPI_MODULE_ID, &provider_module_id, sizeof(GUID));
+        if (error != ERROR_SUCCESS) {
+            result = win32_error_code_to_ebpf_result(error);
+            goto Done;
+        }
+
+        service_path = service_path + service_name.c_str();
+        result = _load_native_module(service_path, &provider_module_id, &count_of_maps, &count_of_programs);
+        if (result != EBPF_SUCCESS) {
+            goto Done;
+        }
+
+        if (count_of_programs == 0) {
+            result = EBPF_INVALID_OBJECT;
+            goto Done;
+        }
+
+        // Allocate buffer for program and map handles.
+        program_handles = (ebpf_handle_t*)calloc(count_of_programs, sizeof(ebpf_handle_t));
+        if (program_handles == nullptr) {
+            result = EBPF_NO_MEMORY;
+            goto Done;
+        }
+
+        if (count_of_maps > 0) {
+            map_handles = (ebpf_handle_t*)calloc(count_of_maps, sizeof(ebpf_handle_t));
+            if (map_handles == nullptr) {
+                result = EBPF_NO_MEMORY;
+                goto Done;
+            }
+        }
+
+        result = _load_native_programs(
+            &provider_module_id, program_type, count_of_maps, map_handles, count_of_programs, program_handles);
+        if (result != EBPF_SUCCESS) {
+            goto Done;
+        }
+
+        // Programs have been loaded and we have the program and map handles.
+        new_object = new (std::nothrow) ebpf_object_t();
+        if (new_object == nullptr) {
+            result = EBPF_NO_MEMORY;
+            goto Done;
+        }
+
+        result = _initialize_ebpf_object_native(
+            file_name, count_of_maps, map_handles, count_of_programs, program_handles, *new_object);
+        if (result != EBPF_SUCCESS) {
+            goto Done;
+        }
+
+        *object = new_object;
+        _ebpf_objects.emplace_back(*object);
+
+        *program_fd = new_object->programs[0]->fd;
+    } catch (const std::bad_alloc&) {
+        result = EBPF_NO_MEMORY;
+        goto Done;
+    } catch (...) {
+        result = EBPF_FAILED;
+        goto Done;
+    }
+
+Done:
+    if (result != EBPF_SUCCESS) {
+        _clean_up_ebpf_object(new_object);
+        if (map_handles != nullptr) {
```

Comment:
is a similar check for `program_handles` not needed?

Reply:
There is a check for `program_handles` at Line 2390 below